c++/error-handling.md

```diff
@@ -0,0 +1,337 @@
+# Contractual Disappointment in C++
+
+## Introduction
+
+[Handling Disappointment in C++](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0157r0.html)
+observes that run-time errors are just one kind of disappointment
+that [a C++ program](https://eel.is/c++draft/basic.link#def:program) might elicit.
+I argue that all disappointment involves the violation of a contract
+and that the correct strategy for handling a violation depends on two things:
+
+1. the parameters of the contract, and
+1. the usage profile of the program.
+
+## Bugs and Errors
+
+Section 4.2 of [P0709R4, Zero-overhead Deterministic Exceptions: Throwing values](https://wg21.link/p0709r4),
+"Proposed cleanup: Don’t report logic errors using exceptions",
+makes clear the distinction between errors and bugs.
+
+Applying this distinction to the contracts above:
+
+* A bug is the violation of a C++ API Contract or the ISO C++ Standard.
+* An error is the violation of a End User Contract.
+
+However, there is an exception to this rule
+which applies to something called the Test User Contract.
+When testing for contract violations,
+the user wishes for bugs to behave far more like errors.
+(I believe that this fact is important to understanding
+why so much disagreement surrounds the C++ Contracts feature.)
+
+## What Is a Contract?
+
+For our purposes, a contract is an _agreement_ between a _user_, and
+a _provider_ about the run-time behaviour of some or all of a program.
+_Violation_ of a contract is what leads to disappointment.
+
+For the sake of simplicity and focus,
+I will concentrate on the obligations imposed on the user by the contract.
+I believe that extending the principles laid out here
+to obligations imposed on the provider is relatively straight-forward.
+
+## What Are the Parameters of a Contract?
+
+The following set of parameters -- expressed as questions -- are helpful in
+understanding the correct error-handling strategy for a contract.
+
+* Where, and in what form, is the agreement expressed?
+* Who is the provider of the contract?
+* Who is the user of the contract?
+* What are the consequences of user contract violation?
+
+## What Contracts Are There?
+
+In a C++ program, some contracts that matter are as follows.
+
+### End User Contract
+
+The ultimate contract that a program must fulfil is to its end user.
+All other contracts below are in support of this fulfilment.
+
+* agreement: program documentation, possibly including a `--help` option
+  or a 'man' page
+* provider: the program author(s)
+* user: the program user
+* violation: user input is sanitized and ill-formed input is handled through
+  normal control flow (including exception handling)
```

Comment:
This describes the case when program user fails to comply with their obligations (like a precondition). It does not mention the case where the program author fails to deliver their part of the job (like a postcondition).

Reply:
As mentioned [here](https://github.com/johnmcfarlane/papers/pull/51/files#diff-046c70b4ff393727ea22013b061f963e90339ce00073ea06899f02cd54c46ceaR38):

> For the sake of simplicity and focus, I will concentrate on the obligations imposed on the user by the contract.

I'm not sure what to do to make this clearer and easier to dip into. Possible options:

1. Reword/repeat the limitations to this section.
1. Include both _user_ and _provider_ violations under _violation_.
1. Rename _violation_ to _user violation_.
1. Add a _provider violation_ parameter to "What Are the Parameters of a Contract?" and explain this is omitted.
1. Add a _provider violation_ parameter to "What Are the Parameters of a Contract?" and then fill it out.

One problem is that I am concentrating on what is needed to help the user get things right and I want to keep the amount of detail low. So probably 3) is worth considering.

Another problem is that this document talks about what a program author should do when they encounter different types of disappointment. I generally don't think it's good for user code to be testing provider violations, e.g.:

```c++
std::vector v;
ASSERT(v.empty());  // user shouldn't do this
```

OTOH, maybe spelling this out is a good idea!

Reply:
To focus only on precondition-like things seems fine to me.  But the next section, "Test User Contract" seems to be talking about something closer to postconditions. And this is where I get confused (about the scope of this paper).

Reply:
You're right. It's the most hard-to-understand section and I got a lot of it wrong. I've rewritten it. Try to forget you read it yet and see if it makes any sense now.